git: add git colocate command

AngelEzquerra opened the pull request:
This new command lets you turn a non colocated git repo into a colocated repo and back (by passing and --undo flag). The command simply implements the instructions found in https://github.com/jj-vcs/jj/blob/main/docs/git-compatibility.md#converting-a-repo-into-a-co-located-repo.

# Checklist

If applicable:

- [x ] I have updated `CHANGELOG.md`
- [ x] I have updated the documentation (`README.md`, `docs/`, `demos/`)
- [ ] I have updated the config schema (`cli/src/config-schema.json`)
- [x ] I have added/updated tests to cover my changes

glehmann commented on cli/src/commands/git/colocate.rs:
```diff
@@ -0,0 +1,247 @@
+    workspace_command: &mut WorkspaceCommandHelper,
+) -> Result<(), CommandError> {
+    if is_colocated_git_workspace(workspace_command.workspace(), workspace_command.repo()) {
+        return Err(user_error("Repository is already co-located with Git"));
```
It might be better just to print the message and return without error.
jj prints `Nothing changed.` and exits with a 0 return code in many places already.

ilyagr replied:
An example:

https://github.com/jj-vcs/jj/blob/1fa53c6a4f1c776b3e0c79920e9253249197276c/cli/src/commands/git/push.rs#L391

AngelEzquerra replied:
Thanks, I'll change it as you guys suggest.

yuja commented on cli/src/commands/git/colocate.rs:
```diff
@@ -0,0 +1,253 @@
+/// Convert a Jujutsu repository to a co-located Jujutsu/git repository
+#[derive(clap::Args, Clone, Debug)]
+pub struct GitColocateArgs {
+    /// Undo a previous colocate operation, reverting the repository back to a
+    /// non-colocated git repository
+    #[arg(long)]
+    undo: bool,
```
Here's previous discussion.
https://github.com/jj-vcs/jj/pull/4833

I don't feel strongly whether we should add separate commands, but if we add a single command that does reverse things, it's probably better to add a mandatory flag/argument. Something like https://github.com/jj-vcs/jj/pull/4833#issuecomment-2555666007, but with different command name.

AngelEzquerra replied:
I've just pushed an updated version that replaces --undo with --enable/--disable, and when no flag is passed indicates the current co-location state and suggests how to change it. I've also added a few more tests and changed the messages a bit to make them a bit more consistent.

PhilipMetzger commented:
This previously existed as https://github.com/jj-vcs/jj/pull/4833 which imo is a simpler version of this, also you squash all things from review comments into the respective parent commit, as we don't squash and merge here. The first commit message also should mention that this fixes https://github.com/jj-vcs/jj/issues/4624 since it is the ticket for it.

joyously commented:
Is there any impact on existing workspaces or clones? (not sure if there are pointers to .git or defined remotes that would be "not found")

PhilipMetzger commented:
> Sorry, when you say that `you squash all things from review comments into the respective parent commit`, you mean that I should squash the fixes into the original commit or that am doing that and I should not?

You aren't doing it, as the PR now has three commits, when it only should be one (there should be no, PR suggestions commit and the final rework also only should exist while the approach is discussed). Because the project doesn't squash merge commits and only rebases them. 

> Also, are you saying that you'd prefer to merge #4833 rather than this version? I'm sorry I had not seen that existing PR.

I'm pretty sure @cormacrelf won't mind if you resubmit that patch for him which would finally land this.

AngelEzquerra commented:
> > Sorry, when you say that `you squash all things from review comments into the respective parent commit`, you mean that I should squash the fixes into the original commit or that am doing that and I should not?
> 
> You aren't doing it, as the PR now has three commits, when it only should be one (there should be no, PR suggestions commit and the final rework also only should exist while the approach is discussed). Because the project doesn't squash merge commits and only rebases them.

OK. I actually prefer to do what you suggest but I was worried it could cause issues with GitHub's comment tracking. I've just pushed a new version which squashes the 3 commits into one.

> > Also, are you saying that you'd prefer to merge #4833 rather than this version? I'm sorry I had not seen that existing PR.
> 
> I'm pretty sure @cormacrelf won't mind if you resubmit that patch for him which would finally land this.

I don't mind which one of the two PRs is merged, but if 

> > Sorry, when you say that `you squash all things from review comments into the respective parent commit`, you mean that I should squash the fixes into the original commit or that am doing that and I should not?
> 
> You aren't doing it, as the PR now has three commits, when it only should be one (there should be no, PR suggestions commit and the final rework also only should exist while the approach is discussed). Because the project doesn't squash merge commits and only rebases them.

OK. Actually I prefer doing what you suggest (since it is so natural when using jj) but I was worried it might cause problems with GitHub's PR comment tracking. I just squashed all 3 changes, updated the commit message to indicate that it fixes #4624 and pushed this new version.

> > Also, are you saying that you'd prefer to merge #4833 rather than this version? I'm sorry I had not seen that existing PR.
> 
> I'm pretty sure @cormacrelf won't mind if you resubmit that patch for him which would finally land this.

I don't mind which PR is merged as long as one is merged, but if #4833 is preferable I think it'd be best if the original author got it through.

glehmann commented:
> > Also, are you saying that you'd prefer to merge #4833 rather than this version? I'm sorry I had not seen that existing PR.
> 
> I'm pretty sure @cormacrelf won't mind if you resubmit that patch for him which would finally land this.

That PR was never finalized, which is unfortunate, but is something quite common when the contributor switches to something else or doesn't have time to work on it anymore.

This PR also adds a way to de-colocate the repository, that the previous one didn't have.

I there a specific reason to prefer the previous one?

PhilipMetzger commented:
> I there a specific reason to prefer the previous one?

It contains the start for colocated workspaces and is overall simpler by not having the `--enable/--disable` flags, the other PR also was stuck on a bikeshed how the reverse command should be named but that should be deferred until there's a need for it (and a native backend existing). I also suggested resubmitting his work since https://github.com/jj-vcs/jj/pull/4644#issuecomment-2941115943 exists but the previous PR is still essentially a rebase away from landing.

AngelEzquerra commented:
> > I there a specific reason to prefer the previous one?
> 
> It contains the start for colocated workspaces and is overall simpler by not having the `--enable/--disable` flags, the other PR also was stuck on a bikeshed how the reverse command should be named but that should be deferred until there's a need for it (and a native backend existing). I also suggested resubmitting his work since [#4644 (comment)](https://github.com/jj-vcs/jj/pull/4644#issuecomment-2941115943) exists but the previous PR is still essentially a rebase away from landing.

I had not considered the jj workspace / git worktree integration which I guess would be nice. No reason why it could not be added later.

It's true that #4833 is a smaller change than this PR but I guess part of the reason is that  it does not support undoing the co-location. Also, even though I initially proposed a similar interface for colocating a repo as what is proposed by #4833, (i.e. a simple `jj git colocate` command without any arguments) after the discussion in this PR I feel that using an `--enable` / `--disable` flags is the better UI. That should be easy to add to #4833 if the original author is up to it.

Honestly, I'm a little worried that unless someone makes a decision either way, both PRs will get stuck and none of them will land anytime soon.

senekor commented:
In terms of the justification for this, we should consider that colocated repos will likely become the default in the October release. (#7213) So I assume the need for turning a non-colocated repo into a colocated one after the fact will be much smaller. How about the other way around? Will there be lots of people wanting to turn their default-colocated repos into non-colocated ones? I don't know.

senekor commented:
In that case, do we need this command at all if colocated repos become the default?

PhilipMetzger commented:
> In that case, do we need this command at all if colocated repos become the default?

yes, since it is a useful helper in its own, since I assume there will be a bunch of people who'll want to try out the 'native' experience and switch back.

ilyagr commented:
(Non-blocking) Suggestion from @algmyr on IRC:

> ```
> jj git colocation enable
> jj git colocation disable
> jj git colocation status
> ```

glehmann commented on cli/src/commands/git/colocate.rs:
```diff
@@ -0,0 +1,293 @@
+/// Manage Jujutsu repository colocation with Git
+#[derive(clap::Args, Clone, Debug)]
+pub struct GitColocateArgs {
+    /// Enable co-location (convert into a co-located Jujutsu/Git repository)
```
```suggestion
    /// Enable colocation (convert into a colocated Jujutsu/Git repository)
```
just to be consistent with the naming.
There are a lot of `co-located` instances. Just find/replace

AngelEzquerra replied:
I don't mind changing it (and in fact prefer `colocate` to `co-locate`), but I went with `co-locate` instead of `colocate` because I saw it used in the glossary and FAQ. Should the hyphenated version be changed everywhere to the non-hyphenated version?

PhilipMetzger replied:
this already has been done see https://github.com/jj-vcs/jj/pull/7419, you only need to rebase this on top of it

glehmann commented on cli/src/commands/git/colocate.rs:
```diff
@@ -0,0 +1,293 @@
+    // Validate that both flags aren't used together
+    if args.enable && args.disable {
+        return Err(user_error(
+            "Cannot specify both --enable and --disable flags",
```
Prefer a `conflict_with` flag on the option. See https://github.com/jj-vcs/jj/blob/6a1163472ef4f2d85d7b2434402b72cabfcb6a6a/cli/src/commands/metaedit.rs#L82 for an example.

AngelEzquerra commented on cli/src/commands/git/colocate.rs:
```diff
@@ -0,0 +1,283 @@
+use crate::git_util::is_colocated_git_workspace;
+use crate::ui::Ui;
+
+/// Manage Jujutsu repository colocation with Git
```
That is a great idea. What would be difference between --relative and --absolute?

AngelEzquerra commented on CHANGELOG.md:
```diff
@@ -88,6 +88,9 @@ to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).
 
+* The new command `jj git colocate` can convert a non-colocated git repo into
+  a colocated repo and vice-versa (using the `--undo` flag).
+
```
Done

AngelEzquerra commented on cli/src/commands/git/colocate.rs:
```diff
@@ -0,0 +1,283 @@
+    let workspace_root = workspace_command.workspace_root();
+    let dot_jj_path = workspace_root.join(".jj");
+    let jj_repo_path = dot_jj_path.join("repo");
+    let git_store_path = jj_repo_path.join("store").join("git");
```
I think this is already handled by the earlier call to `is_colocated_git_workspace` (which does call get_git_backend). The `colocate --enable` command does show the correct `Error: Cannot colocate a Jujutsu workspace.` message. However, the `colocate --disable` command shows a not so great (but perhaps technically speaking) message: `Repository is already not colocated with Git.` which could probably be improved.

AngelEzquerra commented on CHANGELOG.md:
```diff
@@ -16,6 +16,9 @@ to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).
   as regular files with a `.jjconflict` suffix and JSON contents.
 
+* The new command `jj git colocate` can convert a non-colocated git repo into
+  a colocated repo and vice-versa (using the `--undo` flag).
```
That's true. I'll fix it, sorry!

AngelEzquerra commented on cli/tests/test_git_colocate.rs:
```diff
@@ -0,0 +1,233 @@
+    assert_eq!(read_git_target(&workspace_root), "git");
+
+    // Run colocate command
+    let _output = test_env.run_jj_in(&workspace_root, ["git", "colocate", "--enable"]);
```
I was able to remove a couple of them. I kept those where there is a single `run_jj_in`call in the test function. If you think I should replace those I can do so but it would require adding an extra line to set the work_dir.
The tests declare a `let workspace_root = test_env.env_root().join("repo");` which then is used to do a lot "joins" and checks for the existence of files in the repo. If `TestWorkDir` let me do those operations checks directly (or through a `path` method) then it would make much more sense to rewrite the rest of the tests to use the `test_env.work_dir()`, making the tests much more verbose...
Or maybe there is a simple way to do what I have in mind and I just missed it? :)

yuja commented on cli/src/commands/git/colocate.rs:
```diff
@@ -0,0 +1,304 @@
+        ));
+    }
+
+    if args.enable {
+        enable_repository_colocation(ui, &mut workspace_command)
+    } else if args.disable {
+        disable_repository_colocation(ui, &mut workspace_command)
+    } else {
+        show_status(ui, &mut workspace_command)
+    }
```
nit: Add different command for status output? We generally avoid overloading a single command for mutation and status output. I think it's important here because `jj git colocate --enable|--disable` is a destructive operation, which the user shouldn't run casually.

AngelEzquerra replied:
As suggested by you and others, I've changed the `--enable` and `--disable` flags into separate sub-commands. I've also added a `status` sub-command that must be used to get the current colocation status. Calling `jj git colocate` will just show the help message.

AngelEzquerra commented on cli/src/commands/git/colocate.rs:
```diff
@@ -0,0 +1,307 @@
+    }
+}
+
+fn show_status(
```
OK, I won't split it into separate files on this PR then.

AngelEzquerra commented on cli/src/commands/git/colocate.rs:
```diff
@@ -0,0 +1,307 @@
+    use std::fs;
+
+    if !to.exists() {
+        fs::create_dir_all(to)?;
```
OK. This was a bit harder to do that I initially thought due to my inexperience with Rust. I hope I did what you had in mind...

AngelEzquerra commented on cli/src/commands/git/colocate.rs:
```diff
@@ -0,0 +1,307 @@
+
+/// Recursively copy a directory to handle cross-filesystem moves
+fn copy_dir_recursive(from: &Path, to: &Path) -> std::io::Result<()> {
+    use std::fs;
```
Hopefully I understood what you meant. I moved those functions higher in the file, after the GitColocateCommand enum.

martinvonz replied:
That's probably fine but what I meant was that `use std::fs;` should be moved to the top/file level

AngelEzquerra commented on cli/src/commands/git/colocate.rs:
```diff
@@ -0,0 +1,305 @@
+        .arg("-C")
+        .arg(&dot_git_path)
+        .args(["config", "--unset", "core.bare"])
+        .output();
```
I looked at some of the examples. It seems that the gix based code would be less straightforward than the current code since I'd need to find the core section in the config (checking that the config file exists), navigate to it to set `bare` and then save the config... Can we leave this as it is for now?

AngelEzquerra commented on cli/src/commands/git/colocate.rs:
```diff
@@ -0,0 +1,305 @@
+
+    // Finally, update git HEAD by taking a snapshot which triggers git export
+    // This will update .git/HEAD to point to the working-copy commit's parent
+    workspace_command.maybe_snapshot(ui)?;
```
I added the command at the end of the `test_git_colocation_enable_success` function and I got the following result:
@  qpvuntsm test.user@example.com 2001-02-03 08:05:07 e8849ae1
│  (empty) (no description set)
◆  zzzzzzzz root() 00000000

What do you expect to see?  A reference to git_head() there?
How would I reload the workspace_command? I didn't find an obvious way of doing it.

AngelEzquerra replied:
As we have discussed offline, I am still having trouble with your recommendation. @yuja I've pushed my latest version (even though it has a failing test) for your review. Thanks!

yuja commented on cli/src/commands/git/colocate.rs:
```diff
@@ -0,0 +1,305 @@
+    // Try a rename first, falling back to copy + remove in case of failure
+    if std::fs::rename(from, to).is_err() {
+        // If rename fails, do a recursive copy and delete
+        copy_dir_recursive(from, to).map_err(|e| e.source)?;
```
Perhaps, we'll need to copy the repository unconditionally if `external` repository is specified? Since external Git repository isn't managed solely by `jj`, renaming it may break non-jj stuff.

AngelEzquerra replied:
I assume this is OK as is, so I'll mark this conversation as resolved. Please reopen it if you disagree!

yuja replied:
I would just delete code that doesn't make sense with the current implementation. You can add it back when needed.

yuja commented on cli/src/commands/git/colocate.rs:
```diff
@@ -0,0 +1,305 @@
+    // This should be fine since it does not make sense to only ignore parts of
+    // the .jj directory
+    let jj_gitignore_path = dot_jj_path.join(".gitignore");
+    std::fs::write(&jj_gitignore_path, "/*\n").context(&jj_gitignore_path)?;
```
So I suggest doing that later (after reloading the workspace helper.)

AngelEzquerra replied:
I'm having trouble implementing your suggestions. I added `let mut workspace_command = command.workspace_helper_no_snapshot(ui)?;` after I moved the `.git` folder to the root of the workspace. However, if after that I print `working_copy_shared_with_git()` I get `false` (even though the `.git` folder was properly moved).
Any ideas?

yuja commented on cli/src/commands/git/colocation.rs:
```diff
@@ -0,0 +1,310 @@
+
+    maybe_add_gitignore(&workspace_command)?;
+
+    make_git_repo_non_bare(dot_git_path)?;
```
We should make the repo non-bare first, right? Otherwise the reloaded workspace is still non-colocated.

AngelEzquerra replied:
You are right! That seems to fix the issue. Thanks! I've pushed a new version that hopefully addresses this and your other comments.

PhilipMetzger commented on docs/git-compatibility.md:
```diff
@@ -196,8 +215,8 @@ jj new && jj undo
 
-We may officially support this in the future. If you try this, we would
-appreciate feedback and bug reports.
+    We may officially support this in the future. If you try this, we would
+    appreciate feedback and bug reports.
```
nit: I think you can delete the paragraphs above this, since there's a command now. I only noticed it because the file somehow got auto-formatted

AngelEzquerra replied:
I've removed that command but I have kept the description of the manual process in case somebody wants to know how it works. Maybe once the command has been included in jj for a version or two we could remove that as well.

AngelEzquerra commented on cli/tests/test_git_colocate.rs:
```diff
@@ -0,0 +1,200 @@
+    ");
+    insta::assert_snapshot!(output.stderr, @"
+        Hint: To enable colocation, run: `jj git colocate enable`
+        [EOF]");
```
I ensured that all the `insta::assert_snapshot!` strings are indented in the same way.

yuja commented on cli/src/commands/git/colocation.rs:
```diff
@@ -0,0 +1,325 @@
+    // repository does not track Jujutsu's repository
+    maybe_add_gitignore(&workspace_command)?;
+
+    // Finally, update git HEAD to point to the working-copy commit's parent
+    let wc_commit_id = workspace_command.get_wc_commit_id().unwrap().to_string();
+    std::fs::write(&git_head_path, wc_commit_id + "\n").context(&git_head_path)?;
```
Something like that. (I think you can use `workspace_command.start_transaction()`, etc., though.)

To test the result, insert `jj log --ignore-working-copy` or something to ensure that `git_head()` points to the right revision.

> I don't think I can use that function to revert the HEAD to "ref: refs/heads/main\n" 

Maybe we can pass the root commit to revert `HEAD` to an unborn branch (I don't think the content of the HEAD matters), but it might fail because non-colocated repo wouldn't have a Git index. If it didn't work, we can use `gix` API to update the HEAD.

AngelEzquerra commented on cli/src/commands/git/colocation.rs:
```diff
@@ -0,0 +1,370 @@
+    let wc_commit_id = workspace_command.get_wc_commit_id().unwrap().clone();
+    let mut tx = workspace_command.start_transaction();
+    let wc_commit = tx.repo_mut().store().get_commit(&wc_commit_id)?;
+    git::reset_head(tx.repo_mut(), &wc_commit)?;
```
I resolved the problem by also using gix for the `cmd_git_colocation_enable`. I think it makes sense to be consistent. However I have 2 concerns:

1. `jj git colocate enable` now creates a new revision on top of the current revision
2. There is quite a bit of duplicated code between the new `set_git_head_to_commit_id` and `reset_git_head_to_main_branch` functions.

I don't understand why only `jj git colocate enable` (which uses `set_git_head_to_commit_id`) creates a new revision while `jj git colocate disable` (which uses `reset_git_head_to_main_branch`) since both those functions are so similar.

AngelEzquerra replied:
The main problem I have with the current gix based version compared to the one in which I simply edited the git HEAD file manually is that when I call `colocate enable` a new revision is created and I do not understand why.
Also, I don't know if adding a new operation that can be undone by `jj undo` is a good idea. I don't think the colocation command should be undoable (because you want to be able to undo previous operations _after_ you change the colocation but without undoing the colocation itself).
> I think the disable command should also unset git_head().
Do you mean that after `colocation disable` `jj log` should not show the `git_head()`? I think that makes sense but I'm not sure how to do that. I've tried to set `deref` to `true` on `reset_git_head_to_main` but that did not do the trick...
Sorry if there is an obvious solution but I still don't understand a lot of the details...

joyously replied:
> Also, I don't know if adding a new operation that can be undone by `jj undo` is a good idea. I don't think the colocation command should be undoable

Does every command go in the op log? Does that include `config` and `util gc`? Do we have categories of commands, such as those that affect local repo state that isn't sent to remote, those that affect files in the repo, those that report, those that affect `jj` itself?

yuja replied:
> The main problem I have with the current gix based version compared to the one in which I simply edited the git HEAD file manually is that when I call `colocate enable` a new revision is created and I do not understand why.

You appears to set `HEAD` to `@`, not to `@-`. I don't know why it worked before, though.

> Also, I don't know if adding a new operation that can be undone by `jj undo` is a good idea.

As I said, you'll get an operation that cannot be undone if you run `jj git colocation enable && jj`. The second `jj` would snapshot the change you made with `jj git colocation enable` and didn't record as such.

> > Do you mean that after `colocation disable` `jj log` should not show the `git_head()`?

Yes. Do `repo.set_git_head_target(RefTarget::absent())` or something.

AngelEzquerra replied:
I also fixed the `colocation enable` command to point the git head to the workin copy parent's revision, which also resolved the problem I had with new revisions being created when you enabled colocation.
In addition to that I was able to remove the git_head reference entirely by following your advice, @yuja, thank you!

martinvonz commented:
@AngelEzquerra: It looks like this is pretty close to ready to merge. Do you think you will find time to finish it? No rush, though. I'm mostly just checking that you haven't forgotten about it.

AngelEzquerra commented:
> @AngelEzquerra: It looks like this is pretty close to ready to merge. Do you think you will find time to finish it? No rush, though. I'm mostly just checking that you haven't forgotten about it.

OK, I believe I have fixed all the nits raised by @yuja. I think it can be merged now! @yuja, can you do a final review and let me know if I can merge? Thanks!

yuja commented on cli/src/commands/git/colocation.rs:
```diff
@@ -0,0 +1,343 @@
+
+/// Manage Jujutsu repository colocation with Git
+#[derive(clap::Subcommand, Clone, Debug)]
+pub enum GitColocationCommand {
```
nit: sort subcommands as they are listed in help in definition order.

yuja commented on cli/src/commands/git/colocation.rs:
```diff
@@ -0,0 +1,343 @@
+    ///
+    /// This moves the underlying Git repository that is found inside the .jj
+    /// directory to the root of the Jujutsu workspace. This allows you to
+    /// use Git commands directly in the Jujutsu workspace.
```
nit: we usually add command help to `Args` struct.

yuja commented on cli/src/commands/git/colocation.rs:
```diff
@@ -0,0 +1,343 @@
+        return Err(user_error(
+            "This command cannot be used in a non-main Jujutsu workspace.",
+        ));
+    }
```
Perhaps, this can be extracted to helper function and let `cmd_git_colocation_*()` call it. It's weird that we load the workspace twice.

yuja commented on cli/src/commands/git/colocation.rs:
```diff
@@ -0,0 +1,343 @@
+
+    let workspace_root = workspace_command.workspace_root();
+    let dot_jj_path = workspace_root.join(".jj");
+    let jj_repo_path = dot_jj_path.join("repo");
```
nit: maybe we can use `workspace_command.repo_path()`

yuja commented on cli/src/commands/git/colocation.rs:
```diff
@@ -0,0 +1,343 @@
+    if dot_git_path.exists() {
+        return Err(user_error(
+            "A .git directory already exists in the workspace root. Cannot colocate.",
+        ));
```
Do we need to check this again? Perhaps, we can instead do `fs::rename()` first, which would fail if `.git` directory existed for unknown reasons.

AngelEzquerra replied:
This check is not done before this point. At this point we know that the repo is git backed and not colocated, but there could still be an existing .git folder at the root of the repository. I don't like just letting rename fail because the error message will be very generic. I think this should stay as it is.

yuja commented on cli/src/commands/git/colocation.rs:
```diff
@@ -0,0 +1,343 @@
+    // Move the git repository from .jj/repo/store/git to .git
+    if let Err(e) = std::fs::rename(&git_store_path, &dot_git_path) {
+        // Attempt to delete git_target_path if move fails and show an error message
+        let _ = std::fs::remove_file(&git_target_path);
```
Better to leave files on unexpected error. We can do rename first and update the pointer file. It's relatively easy for user to recover from rename error than restoring the original `git_target` file.

yuja commented on cli/src/commands/git/colocation.rs:
```diff
@@ -0,0 +1,343 @@
+    Ok(())
+}
+
+fn make_git_repo_bare(dot_git_path: &std::path::PathBuf) -> Result<(), CommandError> {
```
nit: `&Path` instead of `PathBuf`

yuja commented on cli/src/commands/git/colocation.rs:
```diff
@@ -0,0 +1,334 @@
+
+    // Make sure that the repository supports git colocation commands
+    // before running any subcommand
+    repo_supports_git_colocation_commands(workspace_command)?;
```
I meant this can be moved to `cmd_git_colocation_*()` where `workspace_command` is loaded again.

martinvonz commented on cli/src/commands/git/colocation.rs:
```diff
@@ -0,0 +1,337 @@
+    let wc_commit = workspace_command
+        .repo()
+        .store()
+        .get_commit(workspace_command.get_wc_commit_id().unwrap())?;
```
nit: We shouldn't crash if you run this command without a working copy. To fix:
```rust
let wc_commit_id = workspace_command.get_wc_commit_id().ok_or_else(|| user_error("This command requires a working copy"))?;
let wc_commit = workspace_command.repo().store().get_commit(wc_commit_id)?;
```

But it would be even better to do the lookup of the `wc_commit` earlier and pass it into this method. That way we won't try to perform the colocation and then fail

martinvonz commented on cli/src/commands/git/colocation.rs:
```diff
@@ -0,0 +1,337 @@
+    if tx.repo().has_changes() {
+        tx.finish(ui, "remove git head reference")?;
+    } else {
+        drop(tx);
```
nit: I think you can delete this. The transaction will be dropped right after anyway, right? Same in `set_git_head_to_wc_parent()`

AngelEzquerra replied:
OK, I did not know that simply not finishing the transaction was equivalent to dropping it :)

martinvonz commented on cli/src/commands/git/colocation.rs:
```diff
@@ -0,0 +1,343 @@
+        .store()
+        .get_commit(workspace_command.get_wc_commit_id().unwrap())?;
+    let parent_id = wc_commit.parent_ids()[0].clone();
+    set_git_head_to_commit_id(workspace_command, &parent_id)?;
```
> I see. Actually that does not work. For some reason git_head() is not set properly when I do that.

What happens instead? Tests passed for me. Are we missing some test coverage?

martinvonz commented on cli/src/commands/git/colocation.rs:
```diff
@@ -0,0 +1,336 @@
+    // Move the Git repository from .git into .jj/repo/store/git
+    std::fs::rename(&dot_git_path, &git_store_path).map_err(|e| {
+        user_error_with_message("Failed to move Git repository to .jj/repo/store/git", e)
+    })?;
```
Same thing about moving the `git_store_path.exists()` handling here

AngelEzquerra replied:
`-C` does work fine in my manual tests.

martinvonz commented on cli/src/commands/git/colocation.rs:
```diff
@@ -0,0 +1,336 @@
+    // Remove the .jj/.gitignore file if it exists
+    if jj_gitignore_path.exists() {
+        std::fs::remove_file(&jj_gitignore_path).context(&jj_gitignore_path)?;
+    }
```
We can probably do `std::fs::remove_file(&jj_gitignore_path).ok();` here instead. That would silently ignore any errors to remove the file but that seems pretty harmless.

yuja commented on cli/tests/test_git_colocation.rs:
```diff
@@ -0,0 +1,233 @@
+    );
+    assert_eq!(read_git_target(&workspace_root), "git");
+    assert!(!workspace_root.join(".jj").join(".gitignore").exists());
+}
```
nit: Can you insert `log --ignore-working-copy` snapshot to ensure that `git_head()` is removed?

AngelEzquerra replied:
I had to add a call to jj new and I also added a check that git_head() is set at the beginning of that tests as well.

yuja commented on cli/tests/test_git_colocation.rs:
```diff
@@ -0,0 +1,233 @@
+            .stderr
+            .raw()
+            .contains("Repository is already not colocated with Git.")
+    );
```
nit: just use `insta::assert_snapshot!()`

joyously commented on cli/tests/test_git_colocation.rs:
```diff
@@ -0,0 +1,233 @@
+    let test_env = TestEnvironment::default();
+
+    // Initialize a regular jj repo
+    let _ = test_env.run_jj_in(test_env.env_root(), ["git", "init", "repo"]);
```
Does this assume the default is not colocated? Didn't that change recently?

AngelEzquerra commented on cli/src/commands/git/colocation.rs:
```diff
@@ -0,0 +1,336 @@
+
+    // Update the git_target file to point to the new location of the git repo
+    let git_target_content = "../../../.git";
+    std::fs::write(&git_target_path, git_target_content).context(git_target_content)?;
```
You are right. I have fixed it.

joyously commented on cli/tests/test_git_colocation.rs:
```diff
@@ -0,0 +1,233 @@
+    let _ = test_env.run_jj_in(&workspace_root, ["git", "colocation", "disable"]);
+    assert!(!workspace_root.join(".git").exists());
+    assert_eq!(read_git_target(&workspace_root), "git");
+}
```
I'm curious: is this command "undo-able"?

yuja commented on cli/tests/test_git_colocation.rs:
```diff
@@ -0,0 +1,255 @@
+    let test_env = TestEnvironment::default();
+
+    // Initialize a non-colocated Jujutsu/Git repo
+    let _ = test_env
```
nit: you should delete `let _ = ` so the compiler would emit warning without `.success()`.
